### ISSUE TITLE
Knora freezes / stalls  (without error) during import-tests

### ISSUE THREAD

**Issue by subotic** [dhlab-basel/Knora#945]
As some point in time during [import-tests](https://github.com/dhlab-basel/knora-tests/tree/master/01_import_test), everything stops without any errors thrown.

- GraphDB is still responsive. I can run SPARQL queries from the GraphDB Workbench.
- Sipi is still responsive. I can get an image.
- Knora is not responding to any requests.

**Comment by subotic** [comment]
Changed the import script, so that now no images are used. After 8612 letters now a bit different error message:

```
[info] ----------------------------------------------------------------
[info] Knora API Server started at http://0.0.0.0:3333
[info] ----------------------------------------------------------------
[info] KnoraService - startupTaskRunner finished
[info] [2018-07-19 23:17:24,962] ERROR - KnoraService(akka://webapi) - Unable to run route /v1/values
[info] akka.pattern.AskTimeoutException: Ask timed out on [ActorSelection[Anchor(akka://webapi/), Path(/user/responderManager)]] after [60000 ms]. Sender[null] sent message of type "org.knora.webapi.messages.v1.responder.valuemessages.CreateValueRequestV1".
[info] 	at akka.pattern.PromiseActorRef$.$anonfun$defaultOnTimeout$1(AskSupport.scala:596)
[info] 	at akka.pattern.PromiseActorRef$.$anonfun$apply$1(AskSupport.scala:606)
[info] 	at akka.actor.Scheduler$$anon$4.run(Scheduler.scala:205)
[info] 	at scala.concurrent.Future$InternalCallbackExecutor$.unbatchedExecute(Future.scala:866)
[info] 	at scala.concurrent.BatchingExecutor.execute(BatchingExecutor.scala:109)
[info] 	at scala.concurrent.BatchingExecutor.execute$(BatchingExecutor.scala:103)
[info] 	at scala.concurrent.Future$InternalCallbackExecutor$.execute(Future.scala:864)
[info] 	at akka.actor.LightArrayRevolverScheduler$TaskHolder.executeTask(LightArrayRevolverScheduler.scala:328)
[info] 	at akka.actor.LightArrayRevolverScheduler$$anon$4.executeBucket$1(LightArrayRevolverScheduler.scala:279)
[info] 	at akka.actor.LightArrayRevolverScheduler$$anon$4.nextTick(LightArrayRevolverScheduler.scala:283)
[info] 	at akka.actor.LightArrayRevolverScheduler$$anon$4.run(LightArrayRevolverScheduler.scala:235)
[info] 	at java.lang.Thread.run(Thread.java:748)
[info] [2018-07-19 23:22:25,020] ERROR - Dispatcher - Ask timed out on [ActorSelection[Anchor(akka://webapi/), Path(/user/storeManager)]] after [360000 ms]. Sender[Actor[akka://webapi/user/responderManager/permissionsRouterADM/$k#1470084813]] sent message of type "org.knora.webapi.messages.store.triplestoremessages.SparqlSelectRequest".
[info] akka.pattern.AskTimeoutException: Ask timed out on [ActorSelection[Anchor(akka://webapi/), Path(/user/storeManager)]] after [360000 ms]. Sender[Actor[akka://webapi/user/responderManager/permissionsRouterADM/$k#1470084813]] sent message of type "org.knora.webapi.messages.store.triplestoremessages.SparqlSelectRequest".
[info] 	at akka.pattern.PromiseActorRef$.$anonfun$defaultOnTimeout$1(AskSupport.scala:596)
[info] 	at akka.pattern.PromiseActorRef$.$anonfun$apply$1(AskSupport.scala:606)
[info] 	at akka.actor.Scheduler$$anon$4.run(Scheduler.scala:205)
[info] 	at scala.concurrent.Future$InternalCallbackExecutor$.unbatchedExecute(Future.scala:866)
[info] 	at scala.concurrent.BatchingExecutor.execute(BatchingExecutor.scala:109)
[info] 	at scala.concurrent.BatchingExecutor.execute$(BatchingExecutor.scala:103)
[info] 	at scala.concurrent.Future$InternalCallbackExecutor$.execute(Future.scala:864)
[info] 	at akka.actor.LightArrayRevolverScheduler$TaskHolder.executeTask(LightArrayRevolverScheduler.scala:328)
[info] 	at akka.actor.LightArrayRevolverScheduler$$anon$4.executeBucket$1(LightArrayRevolverScheduler.scala:279)
[info] 	at akka.actor.LightArrayRevolverScheduler$$anon$4.nextTick(LightArrayRevolverScheduler.scala:283)
[info] 	at akka.actor.LightArrayRevolverScheduler$$anon$4.run(LightArrayRevolverScheduler.scala:235)
[info] 	at java.lang.Thread.run(Thread.java:748)
[info] [2018-07-19 23:22:25,021] ERROR - Dispatcher - Ask timed out on [ActorSelection[Anchor(akka://webapi/), Path(/user/responderManager)]] after [360000 ms]. Sender[Actor[akka://webapi/user/responderManager/valuesRouterV1/$a#-1649692540]] sent message of type "org.knora.webapi.messages.admin.responder.permissionsmessages.DefaultObjectAccessPermissionsStringForPropertyGetADM".
[info] akka.pattern.AskTimeoutException: Ask timed out on [ActorSelection[Anchor(akka://webapi/), Path(/user/responderManager)]] after [360000 ms]. Sender[Actor[akka://webapi/user/responderManager/valuesRouterV1/$a#-1649692540]] sent message of type "org.knora.webapi.messages.admin.responder.permissionsmessages.DefaultObjectAccessPermissionsStringForPropertyGetADM".
[info] 	at akka.pattern.PromiseActorRef$.$anonfun$defaultOnTimeout$1(AskSupport.scala:596)
[info] 	at akka.pattern.PromiseActorRef$.$anonfun$apply$1(AskSupport.scala:606)
[info] 	at akka.actor.Scheduler$$anon$4.run(Scheduler.scala:205)
[info] 	at scala.concurrent.Future$InternalCallbackExecutor$.unbatchedExecute(Future.scala:866)
[info] 	at scala.concurrent.BatchingExecutor.execute(BatchingExecutor.scala:109)
[info] 	at scala.concurrent.BatchingExecutor.execute$(BatchingExecutor.scala:103)
[info] 	at scala.concurrent.Future$InternalCallbackExecutor$.execute(Future.scala:864)
[info] 	at akka.actor.LightArrayRevolverScheduler$TaskHolder.executeTask(LightArrayRevolverScheduler.scala:328)
[info] 	at akka.actor.LightArrayRevolverScheduler$$anon$4.executeBucket$1(LightArrayRevolverScheduler.scala:279)
[info] 	at akka.actor.LightArrayRevolverScheduler$$anon$4.nextTick(LightArrayRevolverScheduler.scala:283)
[info] 	at akka.actor.LightArrayRevolverScheduler$$anon$4.run(LightArrayRevolverScheduler.scala:235)
[info] 	at java.lang.Thread.run(Thread.java:748)
```

I don't know if this is the same problem or something new.

**Comment by benjamingeer** [comment]
If you skip the previous letters and just create letter number 8612, does it work?

**Comment by benjamingeer** [comment]
It looks like:

1. `ValuesResponderV1` received a `CreateValueRequestV1`.
2. `ValuesResponderV1` sent a `DefaultObjectAccessPermissionsStringForPropertyGetADM` to `PermissionsResponderADM`.
3. `PermissionsResponderADM` sent a `SparqlSelectRequest` to the triplestore responder and never got a response.

**Comment by subotic** [comment]
I thnk that the same letter is created every time in a loop. So there should in theory be no diferrence beetween 8612 or any other beside the IRIs.

I will try to make it fail faster. Waiting 4 hours after every code change is a bit slow for me.

**Comment by subotic** [comment]
Also, the timeouts don't make much sense. 60 seconds timeout on a call which initiates another call having a timeout of 360 seconds.

**Comment by subotic** [comment]
Hmm, now all 20'000 letters got created and I didn't change anything :-(

**Comment by tobiasschweizer** [comment]
I ran my import again and it stops without any error. I tried to restart Knora and got:

```
webapi KnoraService - startupTaskRunner starting
webapi [2018-07-25 11:54:37,169] ERROR - TcpListener - Bind failed for TCP channel on endpoint [/0.0.0.0:3333]
webapi java.net.BindException: Address already in use
webapi 	at sun.nio.ch.Net.bind0(Native Method)
webapi 	at sun.nio.ch.Net.bind(Net.java:433)
webapi 	at sun.nio.ch.Net.bind(Net.java:425)
webapi 	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
webapi 	at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
webapi 	at akka.io.TcpListener.liftedTree1$1(TcpListener.scala:56)
webapi 	at akka.io.TcpListener.<init>(TcpListener.scala:53)
webapi 	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
webapi 	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
webapi 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
webapi 	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
webapi 	at akka.util.Reflect$.instantiate(Reflect.scala:66)
webapi 	at akka.actor.ArgsReflectConstructor.produce(IndirectActorProducer.scala:96)
webapi 	at akka.actor.Props.newActor(Props.scala:212)
webapi 	at akka.actor.ActorCell.newActor(ActorCell.scala:624)
webapi 	at akka.actor.ActorCell.create(ActorCell.scala:650)
webapi 	at akka.actor.ActorCell.invokeAll$1_aroundBody2(ActorCell.scala:523)
webapi 	at akka.actor.ActorCell$AjcClosure3.run(ActorCell.scala:1)
webapi 	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149)
webapi 	at akka.kamon.instrumentation.ActorSystemMessageInstrumentation.$anonfun$aroundSystemMessageInvoke$1(ActorSystemMessageInstrumentation.scala:36)
webapi 	at kamon.Kamon$.withContext(Kamon.scala:120)
webapi 	at akka.kamon.instrumentation.ActorSystemMessageInstrumentation.aroundSystemMessageInvoke(ActorSystemMessageInstrumentation.scala:36)
webapi 	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:515)
webapi 	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:545)
webapi 	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:283)
webapi 	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
webapi 	at kamon.executors.Executors$InstrumentedExecutorService$$anon$7.run(Executors.scala:270)
webapi 	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:44)
webapi 	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
webapi 	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
webapi 	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
webapi 	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by subotic** [comment]
This seems to be a problem with `reStart`. Using `sbt run` works better.

**Comment by tobiasschweizer** [comment]
I retried with `idle-timeout = 90 s` for the client, still getting the same behavior (no errors thrown).

**Comment by benjamingeer** [comment]
@subotic and @tobiasschweizer could you please try the import test with the `wip/apache-http-client` branch and see if it makes any difference?

It's an experiment that changes `HttpTriplestoreConnector` and `SipiResponderV1` to use the the [Apache HTTP Client](https://hc.apache.org/httpcomponents-client-ga/index.html) library instead of the Akka HTTP Client library.

**Comment by benjamingeer** [comment]
I forgot there was already a branch for this, would you like me to merge the changes on `wip/apache-http-client` into #951?

**Comment by tobiasschweizer** [comment]
That sounds good, thanks a lot. @subotic Would that be ok?

**Comment by subotic** [comment]
Please don't. I finally got the tests working again on this branch. I first want to try out if the changes I made have any effect.

**Comment by subotic** [comment]
@benjamingeer If you want, you can checkout [this](https://github.com/dhlab-basel/knora-tests/tree/master/01_import_test) and run the test.

**Comment by subotic** [comment]
I've ran again the test with facsimiles. The problem is with the SipiResponder (at least this time). Knora didn't freeze:

```
[2018-07-26 19:23:32,538] ERROR - Dispatcher - Unknown error: akka.http.impl.engine.client.OutgoingConnectionBlueprint$UnexpectedConnectionClosureException: The http server closed the connection unexpectedly before delivering responses for 1 outstanding requests
[info] org.knora.webapi.SipiException: Unknown error: akka.http.impl.engine.client.OutgoingConnectionBlueprint$UnexpectedConnectionClosureException: The http server closed the connection unexpectedly before delivering responses for 1 outstanding requests
[info] 	at org.knora.webapi.SipiException$.apply(Exceptions.scala:405)
...
```

Basically, Sipi dropped the connection. The Sipi log doesn't show anything abnormal, though.

I will make a test, that only hammers Sipi. I will also try to handle this error a bit more gracefully in Knora. No need to print out the huge stack trace.

**Comment by lrosenth** [comment]
The sipi expert will be back next week....

Von meinem iPhone gesendet

Am 26.07.2018 um 14:10 schrieb Ivan Subotic <notifications@github.com<mailto:notifications@github.com>>:


I've ran again the test with facsimiles. The problem is with the SipiResponder (at least this time). Knora didn't freeze:

[2018-07-26 19:23:32,538] ERROR - Dispatcher - Unknown error: akka.http.impl.engine.client.OutgoingConnectionBlueprint$UnexpectedConnectionClosureException: The http server closed the connection unexpectedly before delivering responses for 1 outstanding requests
[info] org.knora.webapi.SipiException: Unknown error: akka.http.impl.engine.client.OutgoingConnectionBlueprint$UnexpectedConnectionClosureException: The http server closed the connection unexpectedly before delivering responses for 1 outstanding requests
[info]  at org.knora.webapi.SipiException$.apply(Exceptions.scala:405)
...


Basically, Sipi dropped the connection. The Sipi log doesn't show anything abnormal, though.

I will make a test, that only hammers Sipi. I will also try to handle this error a bit more gracefully in Knora. No need to print out the huge stack trace.

**Comment by subotic** [comment]
The problem with sipi was probably due to Docker and the fact that the images were stored inside the container. I changed that, so now any files are stored outside of the container. Also, I changed the size of the TIFFs to 1 MB.

The test script ran this time up to `795`, at which point Knora froze (took around 1 h). No requests are accepted from the outside. Also, I had to reboot my machine, since the port `3333` was still in use.

Now, I'm running the test against the `wip/ apache-http-client` branch, to see if this maybe makes a difference.

**Comment by benjamingeer** [comment]
My other hypothesis is a deadlock in our code, maybe involving `IriLocker` (we could try replacing `IrILocker` with a dummy implementation that doesn't lock).

**Comment by subotic** [comment]
Good idea. Could you maybe do it on the `wip/apache-http-client` branch?

**Comment by benjamingeer** [comment]
Sure. Did your test with that branch already freeze?

**Comment by subotic** [comment]
I'm currently testing it. The test is still running.

**Comment by subotic** [comment]
or maybe better to wait until the test finishes

**Comment by benjamingeer** [comment]
`IriLocker` shouldn't actually allow deadlocks, because it should throw an exception if it can't get a lock within the timeout, but there could always be a bug. :)

**Comment by benjamingeer** [comment]
If it freezes, could you get a thread dump?

[Take thread dumps from a JVM](https://helpx.adobe.com/experience-manager/kb/TakeThreadDump.html)

**Comment by subotic** [comment]
Hmm, it didn't freeze. So the question now is, is it fixed, or are we not exerting enough pressure on Akka to trigger the underlying problem?

Before I merge this into the `945-knora-freeze` branch, I will run an import with 4000 letters.

**Comment by subotic** [comment]
Finally, I've found out what the problem was. We were using the default dispatcher which got clogged up. Adding our own custom dispatchers of the `thread-pool-executor` type fixed the problem.

**Comment by tobiasschweizer** [comment]
Great job!

**Comment by benjamingeer** [comment]
Yaaaaaaaaaaaaaaay!!!!!

**Comment by subotic** [comment]
:-)
